src/runtime_src/core/common/api/xrt_module.cpp

```diff
@@ -637,8 +640,10 @@ class module_elf : public module_impl
         if (auto search = arg2patchers.find(key_string); search != arg2patchers.end())
           search->second.m_ctrlcode_offset.emplace_back(offset);
         else {
-          auto symbol_type = static_cast<patcher::symbol_type>(rela->r_addend);
-          arg2patchers.emplace(std::move(key_string), patcher{ symbol_type, {offset}, buf_type });
+          uint64_t add_end_higher_28bit = (rela->r_addend & 0xFFFFFFF0) >> 4;
+          uint64_t patch_scheme_lower_4bit = rela->r_addend & 0xF; // NOLINT
+          auto symbol_type = static_cast<patcher::symbol_type>(patch_scheme_lower_4bit);
+          arg2patchers.emplace(std::move(key_string), patcher{ symbol_type, {offset}, buf_type, add_end_higher_28bit });
```

Comment:
addend (high 28 bits) do not belong to a symbol, they are distributed into each offset.

Relocation section '.rela.dyn' at offset 0x47e8 contains 12 entries:
 Offset     Info    Type            Sym.Value  Sym. Name + Addend
00001c80  0000010a unrecognized: a       00000000   4 + 785
00001ed8  0000020a unrecognized: a       00000000   3 + 5
00002018  0000030a unrecognized: a       00000000   4 + 505
000021b0  0000040a unrecognized: a       00000000   3 + 5
000022f0  0000050a unrecognized: a       00000000   4 + 285
00002488  0000060a unrecognized: a       00000000   3 + 5
000025c8  0000070a unrecognized: a       00000000   4 + 5
00002760  0000080a unrecognized: a       00000000   3 + 5
000042a0  0000090a unrecognized: a       00000000   5 + 785
00004398  00000a0a unrecognized: a       00000000   5 + 505
00004490  00000b0a unrecognized: a       00000000   5 + 285
00004588  00000c0a unrecognized: a       00000000   5 + 5

So looks to me, we need to expand m_ctrlcode_offset from std::vector<uint64_t> to std::vector<std::pair<uint64_t, uint32_t>>
First element is offset and second is addend

Reply:
Also, instead of naming it add_end or add_end_higher_28bit, I think a better more informative name should be offset_to_bobase. The other offset could be offset_to_patchbuffer.